File: blog2/internal/entity/article.go
package entity

import (
	"time"
)

type ArticleID string

type Article struct {
	ID          ArticleID  `json:"id"`
	Title       string     `json:"title"`
	Published   bool       `json:"published"`
	PublishedAt *time.Time `json:"publishedAt"`
	CreatedAt   time.Time  `json:"createdAt"`
	UpdatedAt   *time.Time `json:"updatedAt"`
	Tags        []Tag      `json:"tags"`
}

func (t *Article) StateString() string {
	if t.Published {
		return "公開中"
	}
	return "ドラフト"
}

func (t *Article) HasTag(tagID TagID) bool {
	for _, t := range t.Tags {
		if t.ID == tagID {
			return true
		}
	}
	return false
}

func (t *Article) GetTagIDs() []TagID {
	tagIDs := make([]TagID, len(t.Tags))
	for i, tag := range t.Tags {
		tagIDs[i] = tag.ID
	}
	return tagIDs
}

File: blog2/_ts/src/entity/article.ts
import { Tag } from "./tag";

export type ArticleId = string;
export interface Article {
    id: ArticleId,
    title: string,
    published: boolean,
    publishedAt: string,
    publishedAtAsDate: Date | undefined,
    tags: Tag[],
}

export function setupArticle(a: Article) {
    try {
        const d = new Date(Date.parse(a.publishedAt));
        a.publishedAtAsDate = d;
    } catch (e: unknown) {
        // Do nothing
    }
}